File: src/Enterprise/Java/CompositeExample.java
import java.util.List;

interface Component {
    int size();
}

final class FileLeaf implements Component {
    private final int bytes;

    FileLeaf(int bytes) {
        this.bytes = bytes;
    }

    public int size() {
        return bytes;
    }
}

final class FolderComposite implements Component {
    private final List<Component> children;

    FolderComposite(Component... children) {
        this.children = List.of(children);
    }

    public int size() {
        return children.stream().mapToInt(Component::size).sum();
    }
}

public final class CompositeExample {
    public static void main(String[] args) {
        Component readme = new FileLeaf(2);
        Component docs = new FolderComposite(new FileLeaf(3), new FileLeaf(5));
        Component root = new FolderComposite(readme, docs);

        System.out.println("leaf=" + readme.size());
        System.out.println("docs=" + docs.size());
        System.out.println("root=" + root.size());
    }
}

File: src/Enterprise/Kotlin/CompositeExample.kt
interface Component {
    fun size(): Int
}

class FileLeaf(private val bytes: Int) : Component {
    override fun size(): Int = bytes
}

class FolderComposite(private vararg val children: Component) : Component {
    override fun size(): Int = children.sumOf { it.size() }
}

fun main() {
    val readme: Component = FileLeaf(2)
    val docs: Component = FolderComposite(FileLeaf(3), FileLeaf(5))
    val root: Component = FolderComposite(readme, docs)

    println("leaf=${readme.size()}")
    println("docs=${docs.size()}")
    println("root=${root.size()}")
}
